bundles/org.eclipse.core.jobs/src/org/eclipse/core/runtime/jobs/ISchedulableOperation.java

```diff
@@ -0,0 +1,39 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Joerg Kubitz.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Joerg Kubitz - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.core.runtime.jobs;
+
+/**
+ * An interface to mark an Operation that needs an {@link ISchedulingRule}.
+ *
+ * @since 3.13
+ *
+ */
+public interface ISchedulableOperation {
+	/**
+	 * @return an {@link ISchedulingRule} that the operation will acquire in the
+	 *         current thread - if any. To make sure the operation can proceed
```

Comment:
The new javadoc is way better compared to the old one.
Still I struggle with this: `To make sure the operation can proceed without waiting for a blocking thread ... caller should call beginRule()`

The problem is - this is the opposite what the API says, because beginRule() API says `If the rule conflicts with another rule currently running in another thread, this method blocks`.

What you **actually** mean is: `to make sure operation can interrupt autobuild if used in context of ...` but that is too much of a unrelated context for *this* API.

Here is my attempt to fix that (you can copy to clipboard and use Team->Apply patch in Eclipse):
```
diff --git a/bundles/org.eclipse.core.jobs/src/org/eclipse/core/runtime/jobs/ISchedulableOperation.java b/bundles/org.eclipse.core.jobs/src/org/eclipse/core/runtime/jobs/ISchedulableOperation.java
index 432ce62..db672d5 100644
--- a/bundles/org.eclipse.core.jobs/src/org/eclipse/core/runtime/jobs/ISchedulableOperation.java
+++ b/bundles/org.eclipse.core.jobs/src/org/eclipse/core/runtime/jobs/ISchedulableOperation.java
@@ -15,7 +15,8 @@
 
+import org.eclipse.core.runtime.IProgressMonitor;
+
 /**
- * An interface to mark an Operation that needs an {@link ISchedulingRule}.
+ * An interface to mark an operation that needs an {@link ISchedulingRule}.
  *
  * @since 3.13
- *
  */
@@ -23,12 +24,14 @@
 	/**
+	 * Gives the caller a hint if this operation may require a rule to proceed. The
+	 * caller should call
+	 * {@link IJobManager#beginRule(ISchedulingRule, IProgressMonitor)} before the
+	 * operation and {@link IJobManager#endRule(ISchedulingRule)} after if the rule
+	 * is specified.
+	 *
 	 * @return an {@link ISchedulingRule} that the operation will acquire in the
-	 *         current thread - if any. To make sure the operation can proceed
-	 *         without waiting for a blocking thread a caller should call
-	 *         {@link IJobManager#beginRule(ISchedulingRule, org.eclipse.core.runtime.IProgressMonitor)}
-	 *         before the operation and {@link IJobManager#endRule(ISchedulingRule)}
-	 *         after. Returns {@code null} if no rule needed - in that case the
-	 *         caller should not call neither <code>beginRule</code> nor
-	 *         <code>endRule</code>. As this method returns only a hint the
+	 *         current thread - if any. Returns {@code null} if no rule needed - in
+	 *         that case the caller should not call neither <code>beginRule</code>
+	 *         nor <code>endRule</code>. As this method returns only a hint the
 	 *         operation can not assume that the caller already acquired the rule.
-	 *         The Operation still has to acquire it - which will lead to a nested
+	 *         The operation still has to acquire it - which will lead to a nested
 	 *         rule.
```

Reply:
> The new javadoc is way better compared to the old one. Still I struggle with this: `To make sure the operation can proceed without waiting for a blocking thread ... caller should call beginRule()`
> 
> The problem is - this is the opposite what the API says, because beginRule() API says `If the rule conflicts with another rule currently running in another thread, this method blocks`.

i dont't see the problem. the trick here is that when beginRule() has been called **in the same thread** before the operation, then the beginRule in the operation is just nested and guaranteed to be wait free.

Reply:
> i dont't see the problem

But I see it. 

> the trick here is that when beginRule() has been called in the same thread before the operation, then the beginRule in the operation is just nested and guaranteed to be wait free.

This is not correct.

"wait free" can only happen if there is no conflicting rule set in some other running job: `If the rule conflicts with another rule currently running in another thread, this method blocks`. This is same as for lock.aquire(). Sure, if current thread already acquired the lock, it is wait free. 

Here we don't know in which thread what rule is used, so we can't assume that we are running the code with the rule already started and no one else is running with that rule.

I would recommend to use the proposed javadoc or provide another version if you like.

Reply:
i still don't get the point, let's figure out if we are talking about the same:

I state that:
```
// Caller:
beginRule(myRule,null); // 1. may block until conflicting rule ends
{ // called Operation:
  beginRule(myRule,null); // 2. guaranteed wait free, because same thread
  endRule(myRule);
}
endRule(myRule);
```
and that is what this API is for - to give the Operation the wait free beginRule (at 2.).
And the usage will then even put additional goodies, like interrupt autobuild and cancel dialog (at 1.)

Reply:
> and that is what this API is for

Jörg, you seem to believe that the code here is only for your specific case - it is not, because it is **public API**. We don't know how API will be used once we release that and we can't assume it will be used only a case of nested call, at least here it is very generic one.

It is not called `IOperationForNestedRuleBeginOnly`.

If you want something only for case 2 in your example, this is not right API and actually doesn't belong to this bundle at all because it is tailored for a very specific use case. This is the reason I originally won't place this code here at all.

So if we provide **generic API** in jobs bundle that claims to use `beginRule()` to avoid waits it is just wrong message. `beginRule()` is used to acquire a lock, and that is a **potentially blocking** operation. Therefore it is confusing to provide javadoc that says the opposite.

Reply:
ok, i still can not imaging how to abuse the API but it's not important. i updated the javadoc